# Question B

In [17]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

In [18]:
# Import dataset
df = pd.read_csv('exam_dataset.csv')
x = df.iloc[:, [0,1,2,3]].values
y = df.iloc[:, 23].values


# Display the full row to view all attributes
pd.set_option('display.max_columns',500)
df.head()


Age     BusinessTravel  MonthlyIncome  JobSatisfaction  Bonus  \
0   41      Travel_Rarely           5993                4  17979   
1   49  Travel_Frequently           5130                2  20520   
2   37      Travel_Rarely           2090                3   6270   
3   33  Travel_Frequently           2909                3   8727   
4   27      Travel_Rarely           3468                2  10404   

               Department  DistanceFromHome  Education EducationField  \
0                   Sales                 1          2  Life Sciences   
1  Research & Development                 8          1  Life Sciences   
2  Research & Development                 2          2          Other   
3  Research & Development                 3          4  Life Sciences   
4  Research & Development                 2          1        Medical   

   EmployeeCount  EmployeeNumber  EnvironmentSatisfaction  Gender  JobLevel  \
0              1               1                        2  Female         2   
1              1               2                        3    Male         2   
2              1               4                        4    Male         1   
3              1               5                        4  Female         1   
4              1               7                        1    Male         1   

                 JobRole MaritalStatus  PerformanceRating  StockOptionLevel  \
0        Sales Executive        Single                  3                 0   
1     Research Scientist       Married                  4                 1   
2  Laboratory Technician        Single                  3                 0   
3     Research Scientist       Married                  3                 0   
4  Laboratory Technician       Married                  3                 1   

   TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                      0                1               6   
1                      3                3              10   
2                      3                3               0   
3                      3                3               8   
4                      3                3               2   

   YearsSinceLastPromotion OverTime Attrition  
0                        0      Yes       Yes  
1                        1       No        No  
2                        0      Yes       Yes  
3                        3      Yes        No  
4                        2       No        No

In [19]:
# check on the data
d = df['BusinessTravel'].unique()
d.sort()
print(d)

['Non-Travel' 'Travel_Frequently' 'Travel_Rarely']


In [20]:
#use labelencoder to encode categorical data
#in this case BusinessTravel and Attrition column
from sklearn.preprocessing import LabelEncoder
labelencoder_X = LabelEncoder()
x[:, 1] = labelencoder_X.fit_transform(x[:, 1])
print(x)

labelencoder_Y = LabelEncoder()
y = labelencoder_Y.fit_transform(y)
print(y)

[[41 2 5993 4]
 [49 1 5130 2]
 [37 2 2090 3]
 ...
 [27 2 6142 2]
 [49 1 5390 2]
 [34 2 4404 3]]
[1 0 1 ... 0 0 0]


In [21]:
#splitting the dataset into training and test sets, 80:20
from sklearn.model_selection import train_test_split
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.20, random_state = 0)

In [22]:
#standardization because data is in different range
from sklearn.preprocessing import StandardScaler
sc_x = StandardScaler()
x_train = sc_x.fit_transform(x_train)
x_test = sc_x.transform(x_test)

In [23]:
#fitting Naïve Bayes Classifier to the training set
from sklearn.naive_bayes import GaussianNB
classifier= GaussianNB()
classifier.fit(x_train,y_train) # fit X_train dan Y_train into classifier model

GaussianNB()

In [24]:
#predict for test set
y_pred=classifier.predict(x_test)
print(y_pred,y_test)

[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0] [0 0 1 0 1 0 1 0 0 1 0 1 0 1 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0
 0 1 0 0 0 0 0 0 1 0 0 1 1 0 1 0 0 1 0 0 1 0 0 0 0 0 0 0 0 1 1 0 1 0 0 0 0
 0 0 0 0 1 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1 1 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 0 1 1 0 0 0 1 0 0 0 0 0 0 0 0 1 0 0
 1 0 0 0 0 1 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0 0 0 0 0 0 0 1 0
 0 0 0 1 0 0 1 0 0 0 0 0 0 1

In [25]:
#evaluation model using confusion matrix
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
cm

array([[241,   4],
       [ 48,   1]], dtype=int64)

In [26]:
# calculate accuracy of the your prediction from values in cm
from sklearn.metrics import accuracy_score, precision_score, classification_report
accuracy = round(accuracy_score(y_test, y_pred),4)*100
error = round(100 - accuracy,4)

print("Accuracy:",accuracy,'%')
print("Error rate:",error,'%')
print(classification_report(y_test, y_pred))

Accuracy: 82.31 %
Error rate: 17.69 %
              precision    recall  f1-score   support

           0       0.83      0.98      0.90       245
           1       0.20      0.02      0.04        49

    accuracy                           0.82       294
   macro avg       0.52      0.50      0.47       294
weighted avg       0.73      0.82      0.76       294



Quite high accuracy of 82.31% but this does not indicate a good model
Based on the classification report for both predicted values of no (value 0) and yes (value 1),
it indicates that when it predict yes, it is often incorrect
and when it is actually yes, the prediction of yes did not occur oftenly
this is due to the biasness in the dataset whereby only 49 out of 294 test set that
the attrition value is 1 (only 17% in test set)
f1 score of 0.82 is not reliable since the high precision and recall
was biased towards predicted values of no (value 0)